Provide default child_spec/1 for Raft.StateMachine

bitwalker opened the pull request:
- Provides sane default for child_spec/1, but overridable
- Ensures default config for state machine uses a data_dir which is
unique per-node, and per-name

This was one of the things that caught my eye when I was reimplementing a distributed process registry like Swarm on top of `raft` - when running multiple nodes via `iex`, the same default data_dir is used for all nodes and all state machines. In addition, I needed to provide a child_spec for the process, even though there is only one way to run it (start_peer). I implemented more or less this child_spec in my own application, but there is really nothing special about it at all, it should probably be the default and only overridden in those cases where one has a good reason to, rather than forcing everyone to implement the same child_spec in every state machine.

This provides an easy way to add state machines to a supervisor, it's just `{MyModule, []}` and it will automatically be started as a peer under the supervisor on boot. It also ensures that there is no friction with running multiple nodes locally for testing and development, and supports multiple state machines in the same project as well.

I probably should update the docs as part of this PR, but wanted to get your feedback first.

bitwalker committed:
Provide default child_spec/1 for Raft.StateMachine

- Provides sane default for child_spec/1, but overridable
- Ensures default config for state machine uses a data_dir which is
unique per-node, and per-name

keathley commented:
All of this looks really reasonable to me. I had originally not done this because I wasn't sure what api we wanted to use. I wasn't sure if it made more sense to use `start_link` instead of `start_peer`.

Right now we're starting the raft peers (the log process and the raft fsm process) under the `PeerSupervisor`. The `PeerSupervisor` is a child of the raft dynamic supervisor. I've wondered if it makes more sense to start the `PeerSupervisor` in the users supervision tree. We could accommodate that with our `child_spec`. My thinking is that it would give users more control over how they structure their supervision trees, but maybe thats not necessary.

bitwalker commented:
You know what, I had completely overlooked `PeerSupervisor`. In any case, my personal preference would be to have the user manage where in their own supervisor tree to put this stuff, since they may want to deal with failure in their own way, start things conditionally, etc. Not that you can't do that with just calling `Raft.start_peer` wherever, but it's not as clear to me where that would happen, and it loses the benefit of being explicit about how and when things are started and what happens if things crash.

If more than one thing need to be started in concert (as you mentioned, the FSM and the log process), then perhaps the solution is to direct users to do something like so under their own supervisor tree:

```elixir
children = [
  {Raft.PeerSupervisor, [peer: MyRaftFSM, peer_args: args | opts]},
]
```

The assumption is that PeerSupervisor in this case isn't a DynamicSupervisor, but a standard `:one_for_one` or `:one_for_all` supervisor. So one starts a supervisor for the pair (and by pair I mean log/FSM processes), passes in the callback module for the state machine, arguments for that state machine, and any other options for the supervisor (if needed). That spec is just a rough draft, but hopefully you get what I'm going for there.

You get the benefit then of knowing exactly where in your supervisor tree this stuff lives, and how it will deal with failure, and removes that responsibility from `raft`. We could still use the `child_spec` I've defined here, it would just be used now by `Raft.PeerSupervisor` when it constructs the child spec for the FSM process.

keathley commented:
I agree that it would be nice for users to define their own supervision strategies.

What if we provided a `start_link` and `child_spec` in the state machine use macro that defaulted to the `Raft.PeerSupervisor`? If a user had something like:

```elixir
defmodule MyApp.Stack do
  use Raft.StateMachine
end
```

Then they would only need to do:

```elixir
children = [
  MyApp.Stack,
]
```

In their own supervision tree. Under the hood it can start the `Raft.PeerSupervisor` with any arguments that they need to pass in. What do you think? Are we saying the same thing?

bitwalker commented:
Even better, we're on the same page :) - I'll push an update to this that makes that change

bitwalker committed:
Have state machine child spec start peer supervisor

- Ensure stop_peer/1 still works
- Provide sane defaults for peer config

bitwalker commented:
I've updated the PR with the changes we discussed. I think this is nicer for sure. I also added a bit to the README to describe how to set it up under supervision. The only piece there that may need some clarification would be where to introduce `Raft.set_configuration/2` in that setup, but I know configuration is still a bit up in the air.

keathley commented:
This looks great :tada:

I'm looking at the specific test failure and I think I see the bug. I'll open a separate issue for it.

bitwalker commented:
Thanks! Sorry I disappeared for a few days, been a busy week :)
